p2p: Increase inbound capacity for block-relay only connections

mzumsande opened the pull request:
This is joint work with amitiuttarwar.

See issue #28462 for a broader discussion on increasing the number of block-relay-only connections independent of this particular implementation proposal.

We suggest to increase the number of inbound slots allocated to block-relay-only peers by increasing the default maximum connections from 125 to 200, with 50% of inbound slots accessible for tx-relaying peers.
This is a prerequisite for being able to increase the default number of outgoing block-relay-only peers later, because the current inbound capacity of the network is not sufficient.
In order to account for incoming tx-relaying peers separately from incoming block-relay peers, changes to the inbound eviction logic are necessary.

See the next post in this thread for a more detailed explanation and motivation of the changes.

mzumsande committed:
net: increase inbound capacity for block-relay-only connections

..and adjust the eviction logic.
The new default max connection number is 200, the default maximum of tx-relaying
inbounds is limited to 50% of all inbound connections.
With 11 outbound connections, that is (200 - 11) * 0.5 = 94.5.
As a result, the tx-related maximum traffic should not change
drastically.

When we receive an inbound connection and don't have space for another
full-relay peer, we now attempt to evict specifically a full-relay inbound
after receiving the version message of the new peer.

Once this commit is widely deployed, the added inbound capacity will
allow us to increase the number of outgoing block-relay-only connections.

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

mzumsande committed:
net: add options to AttemptToEvictConnection

If the added bool evict_tx_relay_peer_only is set, non-tx-relay
peers will be exempt from disconnection.
Also allow to specify a peer that exempt from eviction.
Both options will be used in later commits.

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

amitiuttarwar committed:
test: add functional test for inbound maxconnection limits

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

amitiuttarwar committed:
init: make inbound tx relay percentage configurable

Permit users to change the amount of inbounds that are permitted to relay
transactions. This is particularly relevant to ensure that superusers that are
not concerned with resource usage are not artificially restricted from offering
many transaction relay slots to the network.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

mzumsande committed:
p2p: trigger possible eviction if we support bloom filters and change a peer to tx relay

Co-authored-by: Amiti Uttarwar <amiti@uttarwar.org>

mzumsande commented:
### Status Quo
We have a default maximum of 125 connections (not counting manual ones). Out of these, 
11 (8 full outbounds, 2 block-relay-only, 1 feeler) are reserved for outbounds, leaving us with 
114 inbound slots. There are currently no limits on how many of the inbounds can support transaction relay.

### Problem
We want to increase the number of inbound slots specifically offered to block-relay-only connections, but don't want to significantly change the number of tx-relaying inbound peers, because we don’t want to radically change the memory and traffic requirements that come with running a full node with the default configuration.
This is complicated by a timing issue: When a new peer connects to us, we don't know yet whether it is block-relay-only before we have processed their version message.
Inbound peer eviction, however, happens right after a peer connects to us and before processing its version msg. This means that we don't yet know during inbound eviction, if the new peer wants tx relay.

### Approach
We propose to introduce a separate maximum of tx-relaying inbound connections, in addition to the existing global maximum.
The new maximum for tx-relaying inbounds gets chosen such that 60% of the total inbound slots can be utilized by tx-relaying peers. We choose a ratio here instead of a fixed number here so that it scales with users changing their `-maxconnections` value from the default. The ratio of 60% is also made adjustable in the form of a new startup parameter. 

### Changes to the eviction logic
Upon a new peer connection, we treat them as tx-relay while doing a possible eviction, until we have processed their version message. Furthermore, we change the eviction logic as follows: 
First check if the maximum for tx-relaying inbounds is exceeded, and if so, run the eviction algorithm specifically for tx-relaying peers (protecting all block-relay-only peers).
If that didn’t result in an eviction, also check if the global maximum is breached, and run the existing eviction algorithm for all peers.
As a consequence of this, block-relay-only peers may occupy some of the tx-relay connection slots like today (no specific limit is imposed for block-relay-only inbounds) but the reverse is not true.

This approach achieves that the number of tx-relaying peers can never exceed the limit without the need to delay the eviction decision after the version processing. One possible downside is that if all tx-relay inbound slots are full but we have capacity for block-relay-only inbound peers, we can evict a tx-relay peer after a new peer connects, even if it’s block-relay-only. However, this affects just one inbound slot: When the incoming peer has sent their version and is known to be block-relay-only, the tx-relay inbound slot it temporarily had will be free again.

### Numbers
Specifically, we propose to increase the total number of connections from 125 to 200, with a  suggested ratio of 60% tx-relay inbounds. With 11 connection slots reserved for outbound peers, 
this results in (200 - 11) * 0.6 = 113 slots for tx-relaying inbound peers, which closely aligns with the current count of 114.
These numbers were chosen to be sufficient to raise the number of block-relay-only connections from 2 to 8 once this patch is widely deployed (see Issue #28462 for more details on this).

### Handling bloom filter peers
The logic gets a bit complicated if a node supports BIP37 (bloom filters), because then an inbound peer can be switched to tx-relay after sending a `FILTERLOAD` or `FILTERCLEAR` message (but never back!), which could result in an excess of tx-relaying inbounds. We propose to solve this by triggering the eviction logic for a tx-relaying from `net_processing` whenever we switch a peer to tx-relay due to a BIP37-related message.

naumenkogs commented:
Concept ACK, specifically on changes to the eviction logic, and handling bloom filters.
Numbers also seem fine I guess.

I was curious whether there is (or will be) any other good reason to delay eviction after the version (apart from making this logic cleaner), but I haven't identified any.

mzumsande commented:
> I was curious whether there is (or will be) any other good reason to delay eviction after the version (apart from making this logic cleaner), but I haven't identified any.

Discussed this with @amitiuttarwar and we think that it would be better to move the tx-related eviction into the version processing logic.
Pros:
- It's conceptually simpler to only evict a tx-relaying peer when we are sure that it is one
- We wouldn't have to evict anyone in the situation where all tx inbound slots are full and a new block-relay inbound connects
- We need the logic for this (`EvictTxPeerIfFull()`) anyway when dealing with bloom-filter peers.

Cons:
- The eviction now happens in two places, before and after the version.

I will update the PR with this soon!

mzumsande commented:
Changed the approach to do the eviction of tx-relaying peers within the version message processing, when we can make use of the `fRelay` flag they sent us - this has the advantage that we don't have to evict new peers before being sure if they want tx relay.

As a result of the later eviction, it became necessary to protect the new peer from being evicted right by `AttemptToEvictConnection()` right after connecting (so that the new peer only gets evicted if no other peer can be found). This isn't done for the existing unconditional inobound eviction in `net` (`CreateNodeFromAcceptedSocket`), because there we would evict a peer before adding the new peer to `m_nodes`.

mzumsande commented on src/net.h:
```diff
@@ -1193,6 +1196,16 @@ class CConnman
     int GetExtraBlockRelayCount() const;
+    /**
+     * If we are at capacity for inbound tx-relay peers, attempt to evict one.
+     * @param[in]   protect_peer    NodeId of a peer we want to protect
```
done

mzumsande commented on src/net.cpp:
```diff
@@ -1700,7 +1700,7 @@ std::pair<size_t, bool> CConnman::SocketSendData(CNode& node) const
  *   simultaneously better at all of them than honest peers.
  */
-bool CConnman::AttemptToEvictConnection(bool evict_tx_relay_peer, std::optional<NodeId> protected_peer)
+bool CConnman::AttemptToEvictConnection(bool evict_tx_relay_peer, std::optional<NodeId> protect_peer)
```
Yes, some changes made it into the wrong commit. Fixed now.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3489,6 +3489,16 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             }
         }
 
+        // If we have too many tx-relaying inbound peers, attempt to evict one (not necessarily this one).
```
Thanks, took your suggestion.

mzumsande commented on src/net.h:
```diff
@@ -1351,7 +1360,13 @@ class CConnman
 
-    bool AttemptToEvictConnection();
+    /**
+     * Try to find an inbound connection to evict when the node is full.
```
True, that is something the caller has to check. I removed the notion of "full" from both `net.h` and `net.cpp`.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3489,6 +3489,16 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
+        // If we have too many tx-relaying inbound peers, attempt to evict one (not necessarily this one).
+        // Only if this fails, disconnect this peer.
+        if (pfrom.IsInboundConn()) {
```
Good point - done.

mzumsande commented on src/net.h:
```diff
@@ -1351,7 +1360,13 @@ class CConnman
+     * @param[in] protect_peer        Protect peer with node id
+     * @return                        True if a node was marked for disconnect
+     */
+    bool AttemptToEvictConnection(bool evict_tx_relay_peer, std::optional<NodeId> protect_peer);
```
changed to use default values.

mzumsande commented on src/net.h:
```diff
@@ -1198,6 +1197,16 @@ class CConnman
+     *                                a peer to evict (all eligible peers are protected)
+     *                                so that the caller can deal with this.
+     */
+    bool EvictTxPeerIfFull(std::optional<NodeId> protect_peer);
```
changed to use default values.

Not sure about the rename suggestion, I think I prefer `EvictTxPeerIfFull`, because I think for naming describing what the function does is more important than describing the return value. Otherwise other spots might call this function in the future with the intention to just perform a check, and then might be surprised about the side effects.

mzumsande commented:
Thanks for the review @amitiuttarwar! i pushed an update, addressing all feedback.

mzumsande commented:
Rebased after #28464 was merged.

0xB10C commented:
I've been running this for about a week now. I currently have 161 inbound connections: 113 (maximum) full-relay connections and 48 block-relay-only connections. 

![image](https://github.com/bitcoin/bitcoin/assets/19157360/eea66aa5-e6ac-40dc-842b-15ca99f49d7f)

0xB10C commented:
In https://github.com/bitcoin/bitcoin/pull/28463#issuecomment-1716393675 you mention that:

>  [we] don't want to **significantly** change the number of tx-relaying inbound peers, because we don’t want to radically change the memory and traffic requirements that come with running a full node with the default configuration.

On multiple full-inbound `master` nodes I'm getting roughly an 80%/20% full-relay/block-relay-only inbound split. This makes sense as we currently open 8 full-relay outbound and 2 block-relay outbound connections (ignoring feelers here). That's roughly 91 full-relay  and 23 block-relay inbound connections. With the new maximum of 113 that an increase of 22 full-relay inbound slots being used and nearly an increase of 25%. I'd argue that this is a significant change.

mzumsande commented:
> I'd argue that this is a significant change.

Hmm yes, in that comment I was thinking about the theoretical limit of all inbound slots being filled with tx-relaying peers. You are right, in practice that amount will be lower because some block-relay-only connections are made to us. In that sense, the introduction of block-relay-only peers likely came with a reduction in bandwidth for popular nodes that were already at capacity before.

One way to accommodate this would be to reduce the percentage, so that there would only be ~90-100 slots for tx-relaying inbound peers instead of 113.

mzumsande commented:
After discussing wit @amitiuttarwar I've now lowered the inbound percentage from 60% to 50% (leading to 95 instead of 113 slots for tx-relaying inbounds). This way, we'll be more aligned with the `typical` number of tx-relaying inbounds today, because nodes today usually have a significant number of block-relay-only inbounds as @0xB10C .
The previous number of 113 was aligned with the `maximum` number of tx-relaying inbounds.

mzumsande commented:
[a9a64f9 ](https://github.com/bitcoin/bitcoin/commit/a9a64f990e1029ed89d2d08e76b96dcba6e8e91e)to [77a612d](https://github.com/bitcoin/bitcoin/commit/77a612dd75eb2383d62d2d21c182684ff2026416): rebased

mzumsande commented:
[6490355 ](https://github.com/bitcoin/bitcoin/commit/6490355409df48fb5bc027126f045b2f5120f814)to [4963e31](https://github.com/bitcoin/bitcoin/commit/4963e31c9a267ba7046a4bae0b17766dce6f5114): Rebased and added `__file__` to functional test (see #30463).

MarnixCroes reviewed (commented):
related doc should also be updated?
https://github.com/bitcoin/bitcoin/blob/c9b7a792e2d8cfe5be14829ee6108d24abf8fcf6/doc/reduce-memory.md?plain=1#L27

https://github.com/bitcoin/bitcoin/blob/c9b7a792e2d8cfe5be14829ee6108d24abf8fcf6/doc/reduce-traffic.md?plain=1#L6

mzumsande committed:
doc: Update docs that refer to -maxconnections

mzumsande commented:
> related doc should also be updated?

Good point, thanks! I added a commit that updates these spots. This would obviously also need release notes, but those could be added at a later point.

naiyoma commented on doc/reduce-traffic.md:
```diff
@@ -3,8 +3,9 @@ Reduce Traffic
-which are outbound. You can therefore, have at most 114 inbound connections.
+By default, Bitcoin Core allows up to 200 connections to different peers, 11 of
+which are outbound. You can therefore, have at most 189 inbound connections, half of
+which can only be taken up by low-traffic block-relay-only peers.
```
nit: I’m trying to understand something. According to the PR description, 50% of inbound slots are allocated for tx-relaying peers, but the commit message seems to imply that 50% are for block-relay-only peers. Does this mean that the inbound slots are equally split between tx-relaying and block-relay-only peers? If so, could the description be expanded for clarity?

mzumsande replied:
What is meant with is that half of the slots can only be accessed by block-relay-only peers, and  the other half is accessible to both:
"with 50% of inbound slots accessible for tx-relaying peers" and "half of
which can **only** be taken up by low-traffic block-relay-only peers" are therefore both saying the same thing:

If we'd have 0 tx-relay inobund peers, in theory all inbound slots could be filled up with block-relay-only peers, whereas if we had 0 block-relay-only inbound peers, still only half of the slots could be filled with tx-relay peers.

naiyoma commented:
Concept ACK to increase the number of maximum connections to **200**, enabling an increase in outgoing block relay connections later on.

mzumsande commented:
Rebased for cmake, and fixed silent conflicts with #30750.

mzumsande commented:
[df1c9c2](https://github.com/bitcoin/bitcoin/commit/df1c9c2e2f4b21578b79b08301bd10279ba0e22f) to [9806bc5](https://github.com/bitcoin/bitcoin/commit/9806bc54b7c053a5ea132a8abe0d27c8f9c5a65c): rebased (there was a silent conflict for `p2p_opportunistic_1p1c.py`, together with the `-maxconnections=94` default limit used in the functional tests)

marcofleon reviewed (commented):
Concept ACK

  Based on the discussion in https://github.com/bitcoin/bitcoin/issues/34542, there seems to be some general agreement that increasing block-relay-only connections is a good first step toward improving partition resistance. I think it makes sense on its own and it would also lay the groundwork for potentially adding reconciliation-based tx relay connections later on.

  I’ll start running fuzzamoto and other relevant fuzz tests on this PR.

instagibbs commented:
concept ACK. I would like some of the erlay discussion to move forward too, because I'm not sure how to judge this all holistically.

mzumsande commented:
[cfa72f8](https://github.com/bitcoin/bitcoin/commit/cfa72f83c1be54454b7759030ef10be26a60ae07) to [0e5c40e](https://github.com/bitcoin/bitcoin/commit/0e5c40e6d22814411286a7186ec1c563064512c9): just rebased, will address other feedback soon

n1bor commented:
I have a node with a lot of inbound connections if you want help testing...

ipv4    ipv6   onion   total   block
in        395     117      16     528
out         7       2       1      10       2
total     402     119      17     538


obviously each time I restart it I lose them all but normally bounces back after a few days.
But should be able to test the 200 limit and that only 94 are for tx nodes.
Let me know what log parameters you would like run with and will deploy this build for a few weeks.

0xB10C commented:
> I ran a previous version of this branch, but probably best to do that again now nearly two years later.

I've been running two nodes with this PR (https://github.com/mzumsande/bitcoin/commit/cfa72f83c1be54454b7759030ef10be26a60ae07) for a while. Running this
```
bitcoin-cli getpeerinfo | jq 'group_by(.connection_type + "/" + (.relaytxes|tostring)) | map({key: .[0].connection_type + " / relaytxes=" + (.[0].relaytxes|tostring), count: length})'
```


Node `charlie` (149 connections total):

```json
[
  {
    "key": "block-relay-only / relaytxes=false",
    "count": 2
  },
  {
    "key": "inbound / relaytxes=false",
    "count": 45
  },
  {
    "key": "inbound / relaytxes=true",
    "count": 94
  },
  {
    "key": "outbound-full-relay / relaytxes=true",
    "count": 8
  }
]
```

Node `ian` (146 connections total)
```json
[
  {
    "key": "block-relay-only / relaytxes=false",
    "count": 2
  },
  {
    "key": "inbound / relaytxes=false",
    "count": 42
  },
  {
    "key": "inbound / relaytxes=true",
    "count": 94
  },
  {
    "key": "outbound-full-relay / relaytxes=true",
    "count": 8
  }
]
```

Compared to a node `bob` that does not run this patch (123 total connections):

```json
[
  {
    "key": "block-relay-only / relaytxes=false",
    "count": 2
  },
  {
    "key": "inbound / relaytxes=false",
    "count": 26
  },
  {
    "key": "inbound / relaytxes=true",
    "count": 87
  },
  {
    "key": "outbound-full-relay / relaytxes=true",
    "count": 8
  }
]
```

instagibbs commented:
will review actively once feedback is addressed, as promised :)

darosior commented:
Concept ACK

brunoerg committed:
test: add test that EvictTxPeerIfFull only evicts tx-relaying peers

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

mzumsande commented on src/net_processing.cpp:
```diff
@@ -5078,6 +5078,11 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
             }
             pfrom.m_bloom_filter_loaded = true;
             pfrom.m_relays_txs = true;
+            if (pfrom.IsInboundConn() && !m_connman.EvictTxPeerIfFull()) {
```
I think it makes sense: the logic in the `VERSION` message follows the existing logic of rather replacing a peer than denying new peers if full (which helps against eclipse attacks).
I don't think that this is important for peers trying to get filters from us - those are likely to be only light clients anyway, and we don't gain much from that connection, so I don't think they need extra protection.

mzumsande commented on src/net.cpp:
```diff
@@ -2486,6 +2492,23 @@ int CConnman::GetExtraBlockRelayCount() const
+        }
+    }
+    if (tx_inbound_peers > m_max_inbound_full_relay) {
+        return AttemptToEvictConnection(/*evict_tx_relay_peer=*/true, protect_peer);
```
Thanks, good find. I added a commit with your suggested test, just adding a short comment why `nServices=0`.

mzumsande commented:
[0e5c40e](https://github.com/bitcoin/bitcoin/commit/0e5c40e6d22814411286a7186ec1c563064512c9) to [6b2fda0](https://github.com/bitcoin/bitcoin/commit/6b2fda077c2c7c126324a3e698c31ab31dad5153): added test by @brunoerg 
[6b2fda0](https://github.com/bitcoin/bitcoin/commit/6b2fda077c2c7c126324a3e698c31ab31dad5153) to [c3b3c7f](https://github.com/bitcoin/bitcoin/commit/c3b3c7fac3d39bd370f390720a6a3c1199b37d88): rebased

Thanks for the reviews and sorry for the delay. I think I have addressed all feedback now!

> I have a node with a lot of inbound connections if you want help testing...

Thanks! Since @0xB10C has been running this for some time now and (to my knowledge) no issues have come up, there is nothing specific that needs testing - though just runing a node with this branch and reporting anything strange will help!

mzumsande commented:
>  I would like some of the erlay discussion to move forward too, because I'm not sure how to judge this all holistically.

My thinking is that Erlay and this PR are two different ways of increasing connectivity, and we should probably only do one at a time, even if they are technically not exclusive.

With Erlay we would also need to increase `-maxconnections` for a few releases before being able to actually add the additional Erlay connections. If we go for more block-relay-only connections instead we would need to do this PR for a few releases and then bump that number.

Erlay connections have the additional benefit of also helping against tx censorship, whereas this PR only strengthens against partitioning/eclipse attacks, so Erlay does more. The downside is the added complexity of Erlay and the added traffic / reconciliation efforts.

instagibbs commented:
@mzumsande we talked offline and the next step for erlay is seeing what reconciliation-only connections would look like. We can move on with this effort for now.

instagibbs commented on src/net.h:
```diff
@@ -1523,7 +1523,13 @@ class CConnman
+     * @param[in] protect_peer        Protect peer with node id
+     * @return                        True if a node was marked for disconnect
+     */
+    bool AttemptToEvictConnection(bool evict_tx_relay_peer = false, std::optional<NodeId> protect_peer = std::nullopt) EXCLUSIVE_LOCKS_REQUIRED(!m_nodes_mutex);
```
e40d62a79c8c38b5d775779c59d4c7877c67ce9c

asking only: might be worth at least making the first arg non-optional?

mzumsande replied:
I did that. I also renamed it to `evict_tx_relay_peer_only` so that it becomes clear that if set to false, a tx peer may still be evicted.

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3738,6 +3738,15 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
         }
 
+        // If we have too many tx-relaying inbound peers, attempt to evict an existing one.
+        // Only if this fails, disconnect this peer.
```
5cdc6415f123b934be294479183bf0297d022494

Given this new callsite which may also result in a `AttemptToEvictConnection`, could one connection cause two disconnects? Perhaps disconnect a blocksonly on the first one, the a tx relay on second? Trying to understand what's possible.

mzumsande replied:
Yes, I think that this would be possible. I'm not sure if it's much of a problem though. 
Since we don't know whether the peer wants tx relay in `CreateNodeFromAcceptedSocket`, I'm not sure how we could avoid it.

instagibbs commented on src/init.cpp:
```diff
@@ -561,7 +561,11 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
     argsman.AddArg("-forcednsseed", strprintf("Always query for peer addresses via DNS lookup (default: %u)", DEFAULT_FORCEDNSSEED), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-listen", strprintf("Accept connections from outside (default: %u if no -proxy, -connect or -maxconnections=0)", DEFAULT_LISTEN), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-listenonion", strprintf("Automatically create Tor onion service (default: %d)", DEFAULT_LISTEN_ONION), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
5cdc6415f123b934be294479183bf0297d022494

in commit message: "As a result, the tx-related maximum traffic should not change
drastically." is there a particular reason we can't shoot for parity?

mzumsande replied:
With "parity", do you mean choose parameters such that the maximum tx-related traffic is exactly the same as now? 

We could do that, but I think having "nice" numbers such as 200 max connections and `FULL_RELAY_INBOUND_PCT=50`  is more useful. 
I am already a bit worried that changing the long-existing number of 125 maximum connections to a mix of two maxima might confuse users. It would be nice if at least the new values would be something humans can remember.

instagibbs commented on src/net.h:
```diff
@@ -1110,6 +1112,7 @@ class CConnman
         m_max_outbound_block_relay = std::min(MAX_BLOCK_RELAY_ONLY_CONNECTIONS, m_max_automatic_connections - m_max_outbound_full_relay);
         m_max_automatic_outbound = m_max_outbound_full_relay + m_max_outbound_block_relay + m_max_feeler;
         m_max_inbound = std::max(0, m_max_automatic_connections - m_max_automatic_outbound);
+        m_max_inbound_full_relay = std::max(0, static_cast<int>(FULL_RELAY_INBOUND_PCT / 100.0 * m_max_inbound));
```
5cdc6415f123b934be294479183bf0297d022494

IIUC at low "maxconnections" values, the int math flooring could result in having no inbound when the argument would otherwise suggest it could have some. Not saying this is worth "fixing" since people generally don't enable listening and reduce the number drastically, just noting

instagibbs reviewed (commented):
initial comments as I dig in a bit

0xB10C commented:
I've been running dbdddbc (from June 4th) for a few weeks and switching the nodes to something else for now (https://github.com/bitcoin/bitcoin/pull/35465#issuecomment-4767022148). Some numbers, similar to https://github.com/bitcoin/bitcoin/pull/28463#issuecomment-4443758694:

On `charlie` (130 connections total):
```
[
  {
    "key": "block-relay-only / relaytxes=false",
    "count": 2
  },
  {
    "key": "inbound / relaytxes=false",
    "count": 26
  },
  {
    "key": "inbound / relaytxes=true",
    "count": 94
  },
  {
    "key": "outbound-full-relay / relaytxes=true",
    "count": 8
  }
]
```

On node `ian` (167 connections total):
```
[
  {
    "key": "block-relay-only / relaytxes=false",
    "count": 2
  },
  {
    "key": "inbound / relaytxes=false",
    "count": 63
  },
  {
    "key": "inbound / relaytxes=true",
    "count": 94
  },
  {
    "key": "outbound-full-relay / relaytxes=true",
    "count": 8
  }
]
```

marcofleon commented on src/net.h:
```diff
@@ -1087,6 +1089,7 @@ class CConnman
     {
         ServiceFlags m_local_services = NODE_NONE;
         int m_max_automatic_connections = 0;
+        int m_full_relay_inbound_percent = 0;
```
I think for fuzz tests like `p2p_handshake` and `process_message` that initialize connman with the testing setup defaults, initializing to 0 for both max connections and the inbound percentage could be an issue?

The tests add peers to `connman.m_nodes` before doing a version handshake, which now calls `EvictTxPeerIfFull()`. Wouldn't it just disconnect a test inbound peer with `m_relay_txs=true` right away?

If they do need to be set to 0 here, then maybe we should override them in `TestingSetup` in `setup_common.cpp`.

instagibbs replied:
Also curious

mzumsande replied:
Good point - I changed both initializations to the defaults, similar to some other variables.

dergoegge reviewed (approved):
Code review ACK c2437c5db1bf1cc647baf3edc5ef67734928938e

instagibbs commented on src/net.cpp:
```diff
@@ -2529,6 +2535,23 @@ int CConnman::GetExtraBlockRelayCount() const
+    {
+        LOCK(m_nodes_mutex);
+        for (const CNode* pnode : m_nodes) {
+            if (pnode->IsInboundConn() && pnode->m_relays_txs) {
```
Match the other functions like GetExtraFullOutboundCount/GetExtraBlockRelayCount?

```suggestion
            if (pnode->IsInboundConn() && !pnode->fDisconnec &&pnode->m_relays_txs) {
```

mzumsande replied:
Done. It makes sense not to count them since `AttemptToEvictConnection` also skips peers with `fDisconnect`.

instagibbs commented on src/init.cpp:
```diff
@@ -578,7 +578,12 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
-    argsman.AddArg("-maxconnections=<n>", strprintf("Maintain at most <n> automatic connections to peers (default: %u). This limit does not apply to connections manually added via -addnode or the addnode RPC, which have a separate limit of %u. It does not apply to short-lived private broadcast connections either, which have a separate limit of %u.", DEFAULT_MAX_PEER_CONNECTIONS, MAX_ADDNODE_CONNE ...
+    argsman.AddArg("-maxconnections=<n>", strprintf("Maintain at most <n> automatic connections to peers (default: %u). %u slots of these are reserved for outgoing connections. See -inboundrelaypercent for more information about limits applied to transaction relay inbound peers. "
+                                                    "This limit does not apply to connections manually added via -addnode or the addnode RPC, which have a separate limit of %u."
+                                                    "It does not apply to short-lived private broadcast connections either, which have a separate limit of %u.",
```
```suggestion
                                                    "It does not apply to short-lived private broadcast connections either, which have a separate limit of %u. ",
```

mzumsande replied:
fixed whitespace

instagibbs commented on src/net_processing.cpp:
```diff
@@ -3772,6 +3772,15 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
 
+        // If we have too many tx-relaying inbound peers, attempt to evict an existing one.
+        // Only if this fails, disconnect this peer.
+        if (pfrom.IsInboundConn() && pfrom.m_relays_txs) {
```
might be confused but: If we're running in blocksonly mode, should we be dropping inbounds because they say it's ok to announce txs to them (but we won't)

mzumsande replied:
not sure about this. `-blocksonly` only means that we don’t want transactions from them, we might send them txns in rare occasions (e.g. via rpc). As long as this is the case, and `TxRelay` is setup with the memory costs for the bloom filter, it seems ok for me to have the same logic as for normal tx-relaying peers.
Besides, there is still the open question if we should make full outbound connections to blocksonly peers (#16418). So in a way, we'd be doing them a favour by dropping them...

instagibbs replied:
Ok I find the memory argument persuasive

instagibbs commented on src/net_processing.cpp:
```diff
@@ -5048,6 +5062,11 @@ void PeerManagerImpl::ProcessMessage(Peer& peer, CNode& pfrom, const std::string
         pfrom.m_bloom_filter_loaded = false;
         pfrom.m_relays_txs = true;
+        if (pfrom.IsInboundConn() && !m_connman.EvictTxPeerIfFull()) {
+            // We don't have room for another tx-relay peer, disconnect
```
nit: could these 3 locations where we're filtering get a helper function to increase grep-ability? ala `MaybeDisconnectForTxRelayCapacity`?

mzumsande replied:
I did that. The resulting function is a bit unintuitive with its return value, and uses the message type so that logging can retrace for which message is was called. Let me know what you think.

instagibbs replied:
lgtm

instagibbs reviewed (commented):
reviewed through c2437c5db1bf1cc647baf3edc5ef67734928938e

mzumsande commented:
[c2437c5](https://github.com/bitcoin/bitcoin/commit/c2437c5db1bf1cc647baf3edc5ef67734928938e) to [c115084](https://github.com/bitcoin/bitcoin/commit/c11508406e6e0cd14c07c58e23085e1a8ff543fa): addressed feedback (I will be quicker in addressing feedback from now on, sorry...)

dergoegge reviewed (approved):
crACK c11508406e6e0cd14c07c58e23085e1a8ff543fa

instagibbs commented:
ACK c11508406e6e0cd14c07c58e23085e1a8ff543fa

marcofleon reviewed (commented):
ACK c11508406e6e0cd14c07c58e23085e1a8ff543fa

achow101 commented:
ACK c11508406e6e0cd14c07c58e23085e1a8ff543fa
